shotover/src/codec/cassandra.rs

```diff
@@ -561,46 +659,106 @@ impl CassandraEncoder {
-                let header_start = dst.len();
+                match compression {
+                    Compression::None => {
+                        // write envelope header with dummy values for those we cant calculate till after we write the message
+                        let header_start = dst.len();
 
-                dst.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
-                let payload_start = dst.len();
+                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
+                        let payload_start = dst.len();
 
-                self.encode_envelope(dst, m)?;
+                        self.encode_envelope(dst, m, Compression::None)?;
 
-                //measure length of message and calculate crc24 and overwrite frame header values
-                let mut payload_len = dst.len() - payload_start;
+                        //measure length of message and calculate crc24 and overwrite frame header values
+                        let mut payload_len = (dst.len() - payload_start) as u64;
 
-                if true {
-                    // TODO if self_contained
-                    payload_len |= 1 << 17;
-                }
+                        if true {
+                            // TODO if self_contained
+                            payload_len |= 1 << 17;
+                        }
 
-                // add header length & header crc
-                let payload_len = &payload_len.to_le_bytes()[..3];
-                dst[header_start..header_start + 3].copy_from_slice(payload_len);
-                dst[header_start + 3..header_start + 6].copy_from_slice(
-                    &cassandra_protocol::crc::crc24(payload_len).to_le_bytes()[..3],
-                );
+                        // add header length & header crc
+                        let payload_len = &payload_len.to_le_bytes()[..3];
+                        dst[header_start..header_start + 3].copy_from_slice(payload_len);
+                        dst[header_start + 3..header_start + 6].copy_from_slice(
+                            &cassandra_protocol::crc::crc24(payload_len).to_le_bytes()[..3],
+                        );
 
-                // add payload crc
-                dst.extend_from_slice(
-                    &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
-                );
+                        // add payload crc
+                        dst.extend_from_slice(
+                            &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
+                        );
+                    }
+                    Compression::Lz4 => {
+                        let header_start = dst.len();
+
+                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
+                        let payload_start = dst.len();
+
+                        let uncompressed_len = self.encode_compressed_payload(dst, m)?;
```

Comment:
we probably need logic to avoid compressing for small messages.
Its a shame that information is lost when we split the frame into separate messages :/ 

We should add a TODO to the tracking issue to investigate this once we have proper benchmarks.
What we have is fine for now.
There are a few possible solutions:
* always compress or never compress
* transform configuration determines if we compress
* reimplement similar logic to the driver for determining if we compress.
* store a `compress: bool` somewhere so we compress messages that came in compressed. (maybe we could store it in the deprecated v4 compression flag :eyes: )

Reply:
Yep, we will also need something similar for non self-contained.

Reply:
https://github.com/shotover/shotover-proxy/issues/160 Added a todo on the issue